enhancements/operator-conditions-lib.md

```diff
@@ -0,0 +1,114 @@
+
+Operators need to be able to read the Condition CR created by OLM, and also have the ability to add/update necessary conditions. This library will provide APIs, using which users can read the Condition CRD present in the cluster for their operator and modify the `.status.conditions` array to add or update any specific condition.
+
+### Goals:
+
+- List the deliverables of the library.
+
+### Non-Goals:
+
+- Discuss on how OLM will process the conditions set in the CRD.
+- List the function signatures/interface types which will be present in the library.
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to add/modify conditions in the `Condition CR`.
+
+#### Story 2
+
+As an operator author, I would like to have the flexibility of being able to read the `Condition CR` and decide to not set any conditions, thereby opting for default OLM behavior.
+
+## Design Details
+
+When an operator is installed in the cluster, OLM will create a `Condition` Custom Resource for the operator. The operator will be able to perform the following actions using the APIs provided in the library:
+
+1. Read the `Condition` CRD:
+
+The library will use controller-runtime's [client][cr_client] to access the resources present in the cluster. To fetch the resource, we will use the [`Get`][cr_get] method of the client, which requires the [`ObjectKey`][cr_objectkey] of type [`types.NamespacedName`][cr_namespacedname]. OLM will set the name and namespace of Condition CR as [environment variables][conditions_followup_pr] while creating the resource. The library will utilize them to obtain the CRD from the cluster.
+
+2. Specify conditions in the CRD:
+
+The operators can introduce OLM-supported conditions which influence OLM's behavior or can also specify any custom (non-OLM supported) condition for communicating the status of the operator to the users. An example Condition CR is given below:
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: Condition
+metadata:
+  name: foo-operator-zdgen-asdf # RandomGen Name
+  namespace: operators
+status:
+   conditions:
+    type: Upgradeable # The name of the `foo` OLM Supported Condition.
+    status: "False"   # The operator is not ready to be upgraded.
+    reason: "migration"
+    message: "The operator is performing a migration."
+    lastTransitionTime: "2020-08-24T23:15:55Z"
+```
+
+The operator will be allowed to only modify the `status` field of condition CR. Operators can either add, delete or update the `status.conditions` array to include OLM-supported or custom conditions. The format and description of the fields present in the conditions can be referred [here][condition_type].
+
+The library will live in [`operator-lib`][operator_lib] repository and the definitions for Condition CRD will be present in [`operator-framwork/api`][api_repo].
```

Comment:
I left a separate comment below, but I think we should remove our status package and direct users to migrate to https://github.com/kubernetes/apimachinery/blob/master/pkg/api/meta/conditions.go

There are not backwards-compat concerns here IMO since operator-lib is not 1.0.